docs/proposals/volumes.md

```diff
@@ -0,0 +1,422 @@
+
+On systems with SELinux enabled, volumes should be relabeled with the correct SELinux context.
+Docker has this capability today; it is desireable for other container runtime implementations to
+provide similar functionality.
+
+Relabeling should be an optional aspect of a volume plugin to accomodate:
+
+1.  volume types for which generalized relabeling support is not sufficient
+2.  testing for each volume plugin individually
+
+#### SELinux and distributed file systems
+
+Some distributed filesystems have complications where SELinux is involved.  Let's look at NFS as an
+example.  Files in NFS volumes cannot have their contexts changed with `chcon` after the volume is
+mounted; the `context` argument must be passed to the `mount` call.
+
+For NFS, there is an SELinux boolean, `virt_use_nfs` that allows containers to use files with the
+SELinux type `nfs_t`.  Setting `virt_use_nfs` on a system would allow containers to use *any* NFS
+mount without the kubelet having to relabel the files.  However, the semantics would be different
+from what might be expected, since SELinux will not provide any protection against containers from
+one pod using files in an NFS mount belonging to another pod.
+
+NFS is a case where the `chcon` approach isn't going to solve the whole problem.  Instead, the NFS
+volume plugin will have to handle calling `mount` with the `context` argument in order to gain
+cross-container protection from SELinux.
+
+Not all distributed file systems support SELinux uniformly.  Some file systems provide only the
+coarse-level control via an SELinux boolean to determine whether containers can use them.  The
+internal API should account for this.
+
+## Proposed Design
+
+Our proposed design should minimize code for handling ownership and SELinux required in:
+
+1.  volume plugins
+2.  the kubelet
+
+### Cluster capability
+
+A new `EnableSELinuxIntegration` capability should be added to the `capabilities.Capabilities`
+struct:
+
+```go
+// Capabilities defines the set of capabilities available within the system.
+// For now these are global.  Eventually they may be per-user
+type Capabilities struct {
+    // descriptions omitted for brevity
+    AllowPrivileged bool
+    HostNetworkSources []string
+    PerConnectionBandwidthLimitBytesPerSec int64
+
+    // EnableSELinuxIntegration controls whether SELinux integration
+    // is expected from Kubernetes
+    EnableSELinuxIntegration bool
+}
+```
+
+There should be validations added to ensure that if SELinux integration is not enabled, pods cannot
+be created with SELinux contexts set.  The Kubelet should be modified so that pods are not started
+if this capability is set and SELinux is not enabled on the node.  Options for how to handle this
+case:
+
+1.  Validate during Kubelet bootstrap and fail the Kubelet start
+2.  Start the Kubelet but report `NotReady` status to the API server
+3.  Start the Kubelet but create events when pods are scheduled, and fail to start pods
```

Comment:
Unless SELinux policy is applied on a per-pod basis, this approach doesn't seem to make sense.